// database/generalDb.js

module.exports = {
  buildGeneralDb({ makeDb, collectionName, ...extraOperations }) {
    return Object.freeze({
      findById,
      insert,
      remove,
      update,
      customFind,
      ...extraOperations
    });

    async function findById ({ _id }) {
      const db = await makeDb();
      const result = await db.collection(collectionName).find({ _id });
      const found = await result.toArray();
      if (found.length === 0) {
        return null;
      }
      return found[0];
    }

    async function insert ({ _id, ...info }) {
      const db = await makeDb();
      const { insertedId } = await db
        .collection(collectionName)
        .insertOne({ _id, ...info });

      return { insertedId };
    }
    async function update ({ _id, ...info }) {
      const db = await makeDb();
      const { matchedCount, modifiedCount } = await db
        .collection(collectionName)
        .updateOne({ _id }, { $set: { ...info } });
      return { matchedCount, modifiedCount };
    }

    async function remove ({ _id }) {
      const db = await makeDb();
      const { deletedCount } = await db.collection(collectionName).deleteOne({ _id });
      return { deletedCount };
    }

    async function customFind (query) {
      const db = await makeDb();
      const result = await db.collection(collectionName).find(query);
      const found = await result.toArray();
      return found;
    }
  }
}
